Documentation: update the documentations related to isolation level and consistency

mitake opened the pull request:
cc @xiang90  @aphyr
I'm working on updating the doc based on our discussion. What I'm mainly trying to do is decoupling the descriptions related to isolation level from the descriptions related to consistency. The definitions of terminologies are still under discussion so I still need to change some parts.

xiang90 commented on Documentation/learning/api_guarantees.md:
```diff
@@ -41,12 +41,16 @@ As with all distributed systems, it is impossible for etcd to ensure [strict con
 #### Isolation
 
-etcd ensures [serializable isolation][serializable_isolation], which is the highest isolation level available in distributed systems. Read operations will never observe any intermediate data.
+etcd ensures [serializable isolation][serializable_isolation], which is the highest isolation level available in transactional database management systems. Read operations will never observe any intermediate data.
```
i think what @aphyr suggest is to simply merge isolation section with consistency section. then we can claim that etcd supports strict serializability.

aphyr commented:
Yeah. Strict serializability has been around for 20 years now; it's a fairly well-known consistency property in the literature, and it's an easy property to explain intuitively!--KyleOn Jan 2, 2020 11:11, Xiang Li <notifications@github.com> wrote:@xiang90 commented on this pull request.



In Documentation/learning/api_guarantees.md:
> @@ -41,12 +41,16 @@ As with all distributed systems, it is impossible for etcd to ensure [strict con
 
 #### Isolation
 
-etcd ensures [serializable isolation][serializable_isolation], which is the highest isolation level available in distributed systems. Read operations will never observe any intermediate data.
+etcd ensures [serializable isolation][serializable_isolation], which is the highest isolation level available in transactional database management systems. Read operations will never observe any intermediate data.

i think what @aphyr suggest is to simply merge isolation section with consistency section. then we can claim that etcd supports strict serializability.

—You are receiving this because you were mentioned.Reply to this email directly, view it on GitHub, or unsubscribe.

mitake commented:
@xiang90 @aphyr ok, then let me reorganize the change. I'll simply merge the sections.

mitake commented:
Updated, I mainly:
- removed some duplicated and conflicting descriptions (`Consistency` section)
- unified the description of isolation level and consistency (former `Linearizability` section)
Could you take a look @xiang90 @aphyr ?

aphyr commented:
This looks better! I suggest two changes:1. You don't... necessarily need to claim serializability is the strongest isolation property. Conflict and strict serializability are "stronger" properties, and depending on who you talk to, I think those could be considered isolation levels too.2. In the example of linearizability, you should define how operations are assigned timestamps: in particular, any timestamp choice is legal between invocation and completion. Another way to phrase this which doesn't need a clock at all is "non-concurrent operations are totally ordered", or "if operation A completes before B begins, A takes effect before B."

xiang90 commented on Documentation/learning/api_guarantees.md:
```diff
@@ -31,36 +31,23 @@ An etcd operation that modifies the key value store is assigned a single increas
-Linearizability (also known as Atomic Consistency or External Consistency) is a consistency level between strict consistency and sequential consistency. 
+etcd ensures linearizability as consistency of replicas basically. As described below, exceptions are watch operations and read operations which explicitly specifies serializable option.
 
-For linearizability, suppose each operation receives a timestamp from a loosely synchronized global clock. Operations are linearized if and only if they always complete as though they were executed in a sequential order and each operation appears to complete in the order specified by the program. Likewise, if an operation’s timestamp precedes another, that operation must also precede the other op ...
```
@aphyr @mitake 

this section was to talk about reality rather than try to redefine "linearizability". The old statement is correct since `loosely synchronized global clock` is strictly weaker than "global wall clock". But tightening the statement is OK I think.

xiang90 commented:
lgtm after accepting @aphyr's suggestion.

aphyr commented:
Er, at the risk of sounding like a broken record: the old version was, in fact, incorrect: it was incompatible both with reality and the commonly accepted definition of linearizability. If you built a system like this, it would fail to be linearizable. It might also, in some cases, be a time machine.

xiang90 commented:
@aphyr The old thing intends to describe what "linearizability" provides to the users of etcd, not what etcd does or what linearizability is. The users of etcd cannot have perfectly synced time, from their point of view it is what can be observed. That is where the confusion comes from.

xiang90 commented:
In another words, we do not intend to define `linearizability ` or how to achieve `linearizability`, but to tell general users what they can expect when etcd is already linearizabile.

aphyr commented:
Again, the definition provided is *incompatible* with linearizability. If etcd is linearizable, it can't also satisfy this property, except in the very particular case where ever clock happens to return timestamps in an order that is identical to the linearization order.

xiang90 commented:
> Again, the definition provided is *incompatible* with linearizability. If etcd is linearizable, it can't also satisfy this property, except in the very particular case where ever clock happens to return timestamps in an order that is identical to the linearization order.

There is just one `global clock` which is `loosely synchronized` to a true wall time, and the client can get time from that single `global clock`. There are no `clocks`.

Yes, if that `loosely sync` is arbitrary (like can go back in time), then even if etcd is linearizable it cannot satisfy the property.

xiang90 commented on Documentation/learning/api_guarantees.md:
```diff
@@ -31,36 +31,23 @@ An etcd operation that modifies the key value store is assigned a single increas
-For linearizability, suppose each operation receives a timestamp from a loosely synchronized global clock. Operations are linearized if and only if they always complete as though they were executed in a sequential order and each operation appears to complete in the order specified by the program. Likewise, if an operation’s timestamp precedes another, that operation must also precede the other op ...
+For linearizability, suppose each operation receives a timestamp from a coordinated time. Operations are linearized if and only if they always complete as though they were executed in a sequential order and each operation appears to complete in the order specified by the program. Likewise, if an operation’s timestamp precedes another, that operation must also precede the other operation in the se ...
 
 For example, consider a client completing a write at time point 1 (*t1*). A client issuing a read at *t2* (for *t2* > *t1*) should receive a value at least as recent as the previous write, completed at *t1*. However, the read might actually complete only by *t3*. Linearizability guarantees the read returns the most current value. Without linearizability guarantee, the returned value, current at * ...
```
@mitake can we add a link to the paper: https://cs.brown.edu/~mph/HerlihyW90/p463-herlihy.pdf?

mitake replied:
got it, I'll add the link of the paper.

xiang90 commented on Documentation/learning/api_guarantees.md:
```diff
@@ -31,36 +31,23 @@ An etcd operation that modifies the key value store is assigned a single increas
+etcd ensures [serializable isolation][serializable_isolation], which is the highest isolation level available in transactional database systems. Read operations will never observe any intermediate data.
 
-Linearizability (also known as Atomic Consistency or External Consistency) is a consistency level between strict consistency and sequential consistency. 
+etcd ensures linearizability as consistency of replicas basically. As described below, exceptions are watch operations and read operations which explicitly specifies serializable option.
```
we can mention that etcd satisfy strict serializable. @aphyr do you know if there is a good definition on `strict serializable` that we can link to?

mitake replied:
yeah, I'll add the link (http://jepsen.io/consistency/models/strict-serializable) provided by @aphyr in addition to Herlihy & Wing

aphyr commented:
Yup! I've got a write-up here (http://jepsen.io/consistency/models/strict-serializable) which cites the original definition from Herlihy & Wing; you could cite that directly instead, if you like!

mitake commented:
@xiang90 @aphyr how about emphasizing that the timestamp is defined as log index of Raft which is defined during appending and commiting? Sometimes log entries (operations) can be overwritten and that's why the original version use the phrase `loosely synchronized`, I guess.

aphyr commented:
I think the entire timestamp discussion is a sort of red herring: "operations being assigned timestamps" is an extra concept which isn't really necessary to define what linearizability is or what etcd does. Defining operation timestamps as Raft log indices only works because the indices of committed Raft operations are *themselves* linearizable. The definition becomes "linearizability means linearizability", which is technically true, but not very helpful!

It doesn't have to be this complicated! The phrasing I use in Jepsen's [consistency model explanations](https://jepsen.io/consistency/models/linearizable) is:

> Linearizability is one of the strongest single-object consistency models, and implies that every operation appears to take place atomically, in some order, consistent with the real-time ordering of those operations: e.g., if operation A completes before operation B begins, then B should logically take effect after A.

Or you could use the way Herlihy and Wing introduce linearizability in their [original paper](http://cs.brown.edu/~mph/HerlihyW90/p463-herlihy.pdf):

> Linearizability provides the illusion that each operation applied by concurrent processes takes effect instantaneously at some point between its invocation and its response.

Either of these sentences captures the essential qualities of linearizability (total order, atomic operations, non-concurrent operations are ordered) without any need for an explicit discussion of clocks. There is, of course, an assumed coordinate time in "invocation" and "response", but coordinate time should not be confused with actual wall clocks, a globally synchronized clock loosely synchronized to coordinate time, or Raft indices, all of which have been discussed in this thread.

xiang90 commented:
> Or you could use the way Herlihy and Wing introduce linearizability in their original paper:

@aphyr agree on this.

mitake commented:
@xiang90 @aphyr sorry for my late response. I updated the branch based on the above discussion. Could you take a look?

xiang90 commented on Documentation/learning/api_guarantees.md:
```diff
@@ -31,36 +31,27 @@ An etcd operation that modifies the key value store is assigned a single increas
+#### Isolation level and consistency of replicas
 
-Linearizability (also known as Atomic Consistency or External Consistency) is a consistency level between strict consistency and sequential consistency. 
+etcd ensures [serializable isolation][serializable_isolation], which is the highest isolation level available in transactional database systems. Read operations will never observe any intermediate data.
```
ensures [strict serializable], which is a strong isolation guarantee for distributed transnational database systems [link to http://jepsen.io/consistency/models/strict-serializable]

mitake replied:
thanks, I updated this part in the latest commit (with slight modification). Could you take a look?

aphyr replied:
Yeah, this looks good.

mitake replied:
Thanks!

xiang90 commented:
lgtm. thanks!

xiang90 commented:
@aphyr it would be great if you can give it another look. thanks!

aphyr commented on Documentation/learning/api_guarantees.md:
```diff
@@ -31,36 +31,24 @@ An etcd operation that modifies the key value store is assigned a single increas
-#### Linearizability
+#### Isolation level and consistency of replicas
+
+etcd ensures [strict serializability][strict_serializability], which is the strongest isolation guarantee of distributed transactional database systems. Read operations will never observe any intermediate data.
```
Technically strict-1SR isn't *the* strongest consistency property, at least by implication--the empty set is the strongest model under that formalism. There are other implemented properties, like strict variants of conflict serializable, which are technically stronger, but I don't think the distinction is all that meaningful for users. I usually say it's "one of the strongest" to hedge my bets. :)

This is almost certainly fine, just noting it in case anyone wants to nitpick later.

aphyr replied:
Ah, now that I'm reading this in context, you might also want to note here that watch ops and "serializable" reads aren't strict serializable, but everything else is.
